Don't cache 4xx/5xx errors permanently for files from stable releases

0t4u opened the pull request:
Wrench's Zig mirror service currently caches 404 errors for files that are expected to be available, like files from releases (see: https://github.com/hexops/mach/issues/1416, https://github.com/mlugg/setup-zig/issues/33).

This PR checks if the requested file belongs to a stable version and has a version number greater than 0.5.0 (see: https://github.com/mlugg/setup-zig#adding-a-mirror), and prevents caching 4xx/5xx errors for those files, since it is expected that these files should be available later.

As a followup, we should also check if the version of the file being requested corresponds to the version string of the latest Zig master or latest Mach nominated version and not cache 4xx/5xx errors for those as well.

A better approach may be to read the index file and ensure all versions present there won't have their 4xx/5xx errors cached.

- [x] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

0t4u committed:
fix: don't cache 4xx/5xx errors for files from stable releases

emidoots commented:
as-is, this change would open a vulnerability which allows anyone to ask the server to spam ziglang.org with requests for versions that are not expected to exist.

A better solution to this would be to continue caching 4xx/5xx errors, but use a cache expiry duration so that the cached response expires after some time, e.g. 15 minutes

0t4u committed:
feat: cache errors with expiry

0t4u commented:
Please let me know if the latest change addresses your concerns, I have set the expiry at 5 minutes for now, which seems like a good amount of time for files to propagate on the upstream

0t4u commented:
@emidoots could you please review again? Thanks

emidoots commented on internal/wrench/http_pkg.go:
```diff
@@ -413,8 +419,13 @@ func (b *Bot) httpPkgEnsureZigDownloadCached(version, versionKind, fname string)
-		fmt.Fprintf(logWriter, "not fetching: %s (cached error %s)\n", url, cachedError)
-		return cachedError
+		if cachedError.expire >= time.Now().Unix() {
+			delete(cachedResponses, url)
```
There's a race condition on this line, we need to hold the mutex while doing this

emidoots commented:
Will merge once that race condition gets fixed, my bad for being so late on the review here!

0t4u committed:
fix: hold mutex during delete
